internal/proxy/powerflex_handler.go

```diff
@@ -893,6 +898,100 @@ func (s *System) volumeUnmapHandler(next http.Handler, enf *quota.RedisEnforceme
 	})
 }
 
+func (s *System) sdcApproveHandler(next http.Handler, sdcapp *sdc.RedisSdcApprover, opaHost string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx, span := trace.SpanFromContext(r.Context()).TracerProvider().Tracer("").Start(r.Context(), "sdcApproveHandler")
+		defer span.End()
+
+		// Read the body.
+		b, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			writeError(w, "powerflex", "failed to read body", http.StatusInternalServerError, s.log)
+			return
+		}
+		defer r.Body.Close()
+
+		var requestBody map[string]json.RawMessage
+		err = json.NewDecoder(bytes.NewReader(b)).Decode(&requestBody)
+		if err != nil {
+			writeError(w, "powerflex", "decoding request body", http.StatusInternalServerError, s.log)
+			return
+		}
+
+		jwtGroup := r.Context().Value(web.JWTTenantName)
+		group, ok := jwtGroup.(string)
+		if !ok {
+			writeError(w, "powerflex", "incorrect type for JWT group", http.StatusInternalServerError, s.log)
+			return
+		}
+
+		jwtValue := r.Context().Value(web.JWTKey)
+		jwtToken, ok := jwtValue.(token.Token)
+		if !ok {
+			writeError(w, "powerflex", "incorrect type for JWT token", http.StatusInternalServerError, s.log)
+			return
+		}
+
+		claims, err := jwtToken.Claims()
+		if err != nil {
+			writeError(w, "powerflex", "decoding token claims", http.StatusInternalServerError, s.log)
+			return
+		}
+
+		s.log.Debugln("Asking OPA...")
+		// Request policy decision from OPA
+		ans, err := decision.Can(func() decision.Query {
+			return decision.Query{
+				Host:   opaHost,
+				Policy: "/karavi/sdc/approve",
+				Input: map[string]interface{}{
+					"claims": claims,
+				},
+			}
+		})
+		if err != nil {
+			s.log.WithError(err).Error("asking OPA for sdc approval decision")
+			writeError(w, "powerflex", fmt.Sprintf("asking OPA for sdc approval decision: %v", err), http.StatusInternalServerError, s.log)
+			return
+		}
+
+		var opaResp OPAResponse
+		err = json.NewDecoder(bytes.NewReader(ans)).Decode(&opaResp)
+		if err != nil {
+			writeError(w, "powerflex", "decoding opa request body", http.StatusInternalServerError, s.log)
+			return
+		}
+		s.log.WithField("opa_response", string(ans)).Debug("OPA Response")
+		if resp := opaResp.Result; !resp.Response.Allowed {
+			switch {
+			case resp.Claims.Group == "":
+				writeError(w, "powerflex", "invalid token", http.StatusUnauthorized, s.log)
+			default:
+				writeError(w, "powerflex", fmt.Sprintf("request denied: %v", resp.Response.Status.Reason), http.StatusBadRequest, s.log)
+			}
+			return
+		}
+
+		qr := sdc.Request{
+			Group: group,
+		}
+		ok, err = sdcapp.CheckSdcApproveFlag(ctx, qr)
+		if err != nil {
+			writeError(w, "powerflex", "sdc approve request failed", http.StatusInternalServerError, s.log)
```

Comment:
We should log the error and send the error in the response.

`writeError(w, "powerflex", fmt.Errorf( "sdc approve request failed: %v", err), http.StatusInternalServerError, s.log)`

Reply:
Since writeError expects an argument of type string, changed it to this:
`writeError(w, "powerflex", fmt.Sprintf("sdc approve request failed: %v", err), http.StatusInternalServerError, s.log)`